### PR TITLE
Delegate tree building to the parser

### PR DESCRIPTION
**Pull request by little-bobby-tables**
Removing the intermediate representation reduces the complexity of the parser and makes it easier to perform line-independent transforms (handling `if-else`, for instance).

I had to add trailing whitespace removal to the preprocessor because I couldn't handle certain edge cases like "blank lines after if-else" (*test/rendering_test.ex*) without severely complicating the grammar.

### CODE REVIEW

**Comment by doomspork** [comment]
Thank you @little-bobby-tables!  @Rakoth and/or myself will take a peek at this and let you know if we have any feedback 😁

File: lib/slime/parser/transform.ex [hunk 1]
```diff
@@ -190,50 +209,28 @@ defmodule Slime.Parser.Transform do
...
-
-    attributes = if @sort_attrs do
-      Enum.sort_by(attributes, fn ({key, _value}) -> key end)
-    else
-      attributes
+    {inline_content, is_closed} = case content do
+      "/" -> {[], true}
+      "" -> {[], false}
+      [] -> {[], false}
+      (%EExNode{} = eex) -> {[eex], false}
```
Comment (Rakoth) [inline lib/slime/parser/transform.ex]: We could use `content` variable here and in next match:
```
%EExNode{} -> {[content], false}
```

File: lib/slime/compiler.ex [hunk 2]
```diff
@@ -3,41 +3,77 @@ defmodule Slime.Compiler do
...
+  def compile(%DoctypeNode{name: name}), do: Doctype.for(name)
+  def compile(%VerbatimTextNode{content: content}), do: compile(content)
+  def compile(%HTMLNode{name: name, spaces: spaces} = tag) do
+    attrs = Enum.map(tag.attributes, &render_attribute/1)
+    tag_head = Enum.join([name | attrs])
+
+    body = cond do
+      tag.closed             -> "<" <> tag_head <> "/>"
+      name in @void_elements -> "<" <> tag_head <> ">"
+      :otherwise             -> "<" <> tag_head <> ">"
```
Comment (Rakoth) [inline lib/slime/compiler.ex]: matter of taste, but I would prefer line-break after `->` in this case

File: lib/slime/compiler.ex [hunk 3]
```diff
@@ -3,41 +3,77 @@ defmodule Slime.Compiler do
...
     keygen source menuitem track wbr
   )
 
-  def compile(tree) do
-    lines_sep = if Application.get_env(:slime, :keep_lines), do: "\n", else: ""
-    tree
-    |> Enum.map(fn branch -> render_branch(branch) end)
-    |> Enum.join(lines_sep)
+  def compile([]), do: ""
+  def compile([h | tree]) do
```
Comment (Rakoth) [inline lib/slime/compiler.ex]: guard will do better here `compile(tags) when is_list(tags)`

File: src/slime_parser.peg.eex [hunk 4]
```diff
@@ -1,32 +1,33 @@
...
-% tags_list <- tag (crlf tag)*;
-% tag <- verbatim_text / space? tag_item (crlf empty_lines:(crlf space?)* indent tags_list dedent)? / '' &.;
-% Also remove dedent from eol definition and modify transforms accordingly
-tags_list <- tag (crlf indent? empty_lines:(space? crlf indent?)* tag:tag dedent*)*;
+tag <- comment / verbatim_text / tag_item;
+tag_item <- space? (embedded_engine / inline_html / code / html_tag);
 
-% NOTE: Leading whitespace is handled separately for comments and verbatim text
-tag <- comment / verbatim_text / space? tag_item / blank:(space? &eol);
+tags <- (tag crlf*)+;
```
Comment (Rakoth) [inline src/slime_parser.peg.eex]: Should we order rules as they appear in grammar, from top to bottom for easier reading?

File: src/slime_parser.peg.eex [hunk 5]
```diff
@@ -1,32 +1,33 @@
...
+nested_tags <- crlf+ indent tags dedent;
 
-tag_item <- (embedded_engine / inline_html / code / html_tag);
+inline_html <- &'<' text_content nested_tags?;
```
Comment (Rakoth) [inline src/slime_parser.peg.eex:11]: Am I missing something, or html with nested children should always have a closing inline html part, so the following examples should raise a parsing error.
```
<img />
  | test
```
```
<div>
  | test
</div>
  | test 1
```

File: src/slime_parser.peg.eex [hunk 6]
```diff
@@ -1,32 +1,33 @@
...
-inline_html <- &'<' text_content;
+html_tag <- inline_tag / simple_tag;
+
+inline_tag <- tag:tag_shortcut spaces:tag_spaces? ':'
+  space children:(inline_tag / simple_tag);
 
-code <- inline:('=' '='? tag_spaces? / '-') space? code:code_lines;
-code_lines <- code_line / code_line_with_brake crlf (indent code_lines dedent / code_lines);
+simple_tag <- tag:tag_shortcut spaces:tag_spaces space?
+  content:attributes_with_content children:nested_tags?;
```
Comment (Rakoth) [inline src/slime_parser.peg.eex]: There are some invalid test cases, currently supported by accident. In ruby-slim tag with inline content could not have nested children tags, (if there are nested lines, they treated as inline content continuation). In some cases it as already not handled by slime, so I think it is time to fix this behaviour, or at least have a parsing error for this case.
For example:
```
div test
  test
```
in slim renders to
```
<div>test
test</div>
```

**Review by Rakoth — COMMENTED** [review]
This is a great work done, @little-bobby-tables ! I left some comments, there are two about parser grammar, which we could open as separate issues, if you think it is not in scope of this changes.

**Comment by Rakoth** [comment]
There is also some problem with code continuation lines:
```
= Enum.join([1,
  2])

p test
```
raises a parser error about unexpected dedent on line 3

File: src/slime_parser.peg.eex (hunk 5 above)
Comment (little-bobby-tables) [reply src/slime_parser.peg.eex:11]: Slim does not check for the closing tag. It's an interesting idea — it could be handled similar to `if-else` — but I think it is out of scope of this PR, yes.

**Comment by little-bobby-tables** [comment]
@Rakoth, I think I've addressed the issues you've brought up. I couldn't pinpoint exactly why the code continuation rule was failing; I believe it was due to some other rule that used to consume dedents past it, but I'm not sure. Either way, since the preprocessor already skips indentation for lines broken by a backslash, I've just added commas to the regular expression there and that seemed to solve it.

I also haven't replied to your comment about grammar reordering. I don't feel strongly about it, it's probably better to do it outside of this changeset.

@doomspork: the PR now includes a breaking change concerning the behavior of nodes with inline content. What used to be rendered as nested tags:
```
h1 Secret:
  = word
```
is now rendered as text, so as to fit this example from the Slim documentation:
```
p No items found. Please add some inventory.
  Thank you!
```
WDYT?

**Review by Rakoth — APPROVED** [review]
👍 LGTM

**Comment by Rakoth** [comment]
@doomspork I think we good to merge this PR, any objections?

**Comment by doomspork** [comment]
No sir @Rakoth!  This looks good to me.

Thank you again @little-bobby-tables, this is great!
